homework/index.js

```diff
@@ -23,24 +10,157 @@
       const value = options[key];
       if (key === 'text') {
         elem.textContent = value;
+      } else if (key === 'html') {
+        elem.innerHTML = value;
       } else {
         elem.setAttribute(key, value);
       }
     });
     return elem;
   }
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-        return;
+  function addRow(tbody, label, value) {
+    const tr = createAndAppend('tr', tbody);
+    createAndAppend('td', tr, { text: `${label}:` });
+    createAndAppend('td', tr, { text: value });
+    return tr;
+  }
+
+  function renderError(error, container) {
+    createAndAppend('div', container, { text: error.message, class: 'alert-error' });
+  }
+
+  function fetchJSON(url) {
+    return new Promise((resolve, reject) => {
+      const xhr = new XMLHttpRequest();
+      xhr.open('GET', url);
+      xhr.responseType = 'json';
+      xhr.onload = () => {
+        if (xhr.status >= 200 && xhr.status <= 299) {
+          resolve(xhr.response);
+        } else {
+          reject(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
+        }
+      };
+      xhr.onerror = () => reject(new Error('Network request failed'));
+      xhr.send();
+    });
+  }
+
+  function showContributions(urlText, container) {
+    const infoRight = createAndAppend('div', container, {
+      class: 'info-right',
+      text: 'Contributions:',
+    });
+
+    function renderContributions(contributors) {
```

Comment:
For the same reason that you should avoid global variables, you should also avoid nested named functions. It is tempting to use them, because you have access to the variables and parameters of the outer function, but again it can become difficult to track how this is all interconnected. You can place this function at the file level by supplying the value it needs from the outer function through the parameter list.

Also shown here is how you can deal with no contributors (`contributors == null`)

```js
function renderContributions(contributors, infoRight) {
  if (contributors == null) {
    return;
  }
  contributors.forEach(contributor => {
  ...
}

Then, where you call your function, supply the extra parameter:

```js
function showContributions(urlText, container) {
  const infoRight = createAndAppend('div', container, {
    class: 'info-right',
    text: 'Contributions:',
  });

  fetchJSON(urlText)
    .then(contributors => renderContributions(contributors, infoRight))  // <- extra param
    .catch(error => renderError(error, infoRight));
}
```